common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -976,9 +977,28 @@ public UTF8String replace(UTF8String search, UTF8String replace) {
     if (EMPTY_UTF8.equals(search)) {
       return this;
     }
-    String replaced = toString().replace(
-      search.toString(), replace.toString());
-    return fromString(replaced);
+    return replace(search.toString(), replace.toString());
+  }
+
+  public UTF8String replace(String search, String replace) {
+    String before = toString();
+    String after;
+    if (search.length() == 1 && replace.length() == 1) {
+      // Use single-character-replacement fast path
+      after = before.replace(search.charAt(0), replace.charAt(0));
+    } else {
+      // In Java 8, String.replace() is implemented using a regex and is therefore
+      // somewhat inefficient (see https://bugs.openjdk.java.net/browse/JDK-8058779).
+      // This is fixed in Java 9, but in Java 8 we can use Commons StringUtils instead:
+      after = StringUtils.replace(before, search, replace);
+    }
+    // Use reference equality to cheaply detect whether the replacement had no effect,
+    // in which case we can simply return the original UTF8String and save some copying.
+    if (before == after) {
```

Comment:
You need `.equals()` here, no? the objects almost certainly won't be the same String?

Reply:
They generally will be the same if no characters were replaced. For example:

```scala
// To avoid worrying about interning of strings, define some random ones at runtime:
@ def randomString() = "ABCDEF" + scala.math.random()
defined function randomString

@ randomString
res9: String = "ABCDEF0.24700533055374918"

@ res9.replace('Z', 'z') eq res9
res10: Boolean = true

@ res9.replace("ZZZ", "z") eq res9
res11: Boolean = true

@ import $ivy.`org.apache.commons:commons-lang3:3.9`, org.apache.commons.lang3.StringUtils
import $ivy.$                                     , org.apache.commons.lang3.StringUtils

@ StringUtils.replace(res9, "ZZZ", "z") eq res9
res14: Boolean = true

// String is technically unchanged, but a different string object:
@ StringUtils.replace(res9, "A", "A") eq res9
res15: Boolean = false
```

Reply:
I see, OK. `String.equals` also checks this immediately. Would it be worth the tiny overhead of the method call to also catch the cases where the implementations don't happen to return the same object but are still the same? don't know.

Reply:
Thinking about this case-by-case, I think there are five replacement outcomes to consider:

1. No replacements are performed because `search` does not match:
  a. Returns original String object
  b. Returns different-but-equivalent String object
2. Replacements are performed because `search` matches:
  a. `search != replacement`, so returns a different String object.
  b. `search == replacement` and returns the original String object.
  c. `search == replacement` and returns a different-but-equivalent String object.

I think (2b) and (2c) are exceedingly unlikely because users probably won't set `search == replacement`. 

Calling `.equals()` therefore only really benefits us in case (1b), which I don't think will occur in practice with most implementations of `StringUtils.replace()` and `String.replace()`.